common/types/overflow.go

```diff
@@ -222,57 +259,99 @@ func subtractTimeChecked(x, y time.Time) (time.Duration, bool) {
 	nsec2 := y.Sub(y.Truncate(time.Second)).Nanoseconds() // Get nanoseconds by truncating and subtracting.
 
 	// Subtract seconds first, detecting any overflow.
-	sec, ok := subtractInt64Checked(sec1, sec2)
-	if !ok {
-		return time.Duration(0), false
+	sec, err := subtractInt64Checked(sec1, sec2)
+	if err != nil {
+		return time.Duration(0), err
 	}
 
 	// Nanoseconds cannot overflow as time.Time normalizes them to [0, 999999999].
 	nsec := nsec1 - nsec2
 
-	// We need to normalize nanoseconds to be positive and carry extra nanoseconds to seconds.
```

Comment:
Note, the nanos are properly normalized in this case, so the logic ported from `time.Unix` was redundant and removed.